[Royce W] ip

Roycecodes opened the pull request:
Level-1
Level-2
Level-3

Roycecodes committed:
removed any duke and renamed to Binla

Roycecodes committed:
added some comments

Roycecodes committed:
Level - 1 added Echo

Roycecodes committed:
put echo into new class

Roycecodes committed:
Level-2 list: modified echo to become list

Roycecodes committed:
Level-3 mark as done

Roycecodes committed:
Add Task Class,Parser Class,Changed Main class name

Having all classes in a single class made code reading difficult

added additional .java files to handle task and parser class

done to ensure code readability and abstraction

Additionally, past methods (markAsDone) is deprecated and needs to be
recoded

Roycecodes committed:
Add support for Done and Undone methods

after including different task classes done and undone methods became
deprecated

updated done and undone methods that directly interact with class

additionally want to add setter methods for changing deadline and event
times

Roycecodes committed:
Update command identification

previously chatbot identified tasks from whether it contains key words
such as /by and /from

now chatbot checks first word to identify command to follow standards in
instructions

roshnidaksha commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+import java.util.Scanner;
+
+public class BinlaDan {
+    static void byeText(){
```
Perhaps a more intuitive method name like printByeText or printExitMessage?

Roycecodes replied:
Edited. Thanks!

roshnidaksha commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+
+public class BinlaDan {
+    static void byeText(){
+        System.out.println("_________________________________________________________________________________________");
```
Should you be using your printLine() method here? I noticed the same issue in the next line too.

roshnidaksha commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+                + "$$ |  $$ |  $$ |  $$ |\\$$$ |$$ |      $$ |  $$ |      $$ |  $$ |$$ |  $$ |$$ |\\$$$ | \n"
+                + "$$$$$$$  |$$$$$$\\ $$ | \\$$ |$$$$$$$$\\ $$ |  $$ |      $$$$$$$  |$$ |  $$ |$$ | \\$$ | \n"
+                + "\\_______/ \\______|\\__|  \\__|\\________|\\__|  \\__|      \\_______/ \\__|  \\__|\\__|  \\__| \n";
+        // this is my chatbot logo
```
Maybe a more descriptive comment? This is redundant. The same issue is seen at several other places as well.

roshnidaksha commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+
+
+}
+class ListState{
```
Should this be extracted into a separate file ListState?

Roycecodes replied:
You're right. Thanks!

roshnidaksha commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+            markAsUndone(Integer.parseInt(taskNumber));
+        }
+        else if(receivedText.indexOf("deadline") == 0){//create a deadlineTask
+            String[] parsedDeadline = Parser.parseDeadline(receivedText);
```
I like how you extracted commands related to parsing into a new class.

Roycecodes replied:
Thank you!

roshnidaksha commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+        String scannedText = "";
+        while(!scannedText.equals("bye")) {     //bye breaks the while loop
+            scannedText = scanner.nextLine();   //scanner scans console for strings
+           if( !scannedText.equals("bye")) {
```
If and while conditions are the same, there is some redundancy here.

Roycecodes replied:
Contacted the logic to fit within the while statement, much neater now. Thanks!

roshnidaksha commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+        }
+
+    }
+    static void listUpdateText(){
```
Perhaps a more intuitive method name that resonates with its function?

Roycecodes replied:
Changed to updateListText. Thanks for the suggession!

roshnidaksha commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,90 @@
+
+
+}
+class Todo extends Task{ // subclass of Task
+    protected boolean done; // to save done value
+    public Todo(String description){
+        // new constructor because constructors are not inherited
+        super(description);
+
+    }
+
+    public boolean isDone(){// done value getter
+        return done;
+    }
+    public void setDone(boolean args) { // done value setter
+        this.done = args;
+    }
+    public String toString() {
+        String addOutput = this.done? "[/]" : "[X]"; //give check if done cross if not
+        return addOutput +" " + super.toString();
+    }
+
+}
```
Should this be extracted into a separate file Todo.java? I noticed the same issued in subsequent lines as well.

roshnidaksha reviewed (commented):
Good effort. Few changes required.

cj143n commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+        }
+        else if(receivedText.indexOf("done") == 0){
+            int indexOfDone = receivedText.indexOf("done");
+            String taskNumber = receivedText.substring(indexOfDone+5).trim();
```
May be better to clarify what the number means

Roycecodes replied:
Nice catch, done!

cj143n commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+        System.out.print("Added new target: ");
+        System.out.println(description); // echos what is added to list
+
+    }
```
May be neater to add a line after this bracket

Roycecodes replied:
not sure what you mean by this one!

cj143n commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,90 @@
+
+
+}
+class Todo extends Task{ // subclass of Task
+    protected boolean done; // to save done value
+    public Todo(String description){
+        // new constructor because constructors are not inherited
+        super(description);
+
+    }
+
+    public boolean isDone(){// done value getter
+        return done;
+    }
+    public void setDone(boolean args) { // done value setter
+        this.done = args;
+    }
+    public String toString() {
+        String addOutput = this.done? "[/]" : "[X]"; //give check if done cross if not
+        return addOutput +" " + super.toString();
+    }
+
+}
+class Deadline extends Todo{ // subclass of Todo
```
may be better to split the other subclasses into other files

Roycecodes replied:
You're right. Thanks!

cj143n commented on src/main/java/BinlaDan.java:
```diff
@@ -0,0 +1,205 @@
+        System.out.println("your glorious list has grown to " + listIndex);
+    }
+
+    static void addTask(String description){
+        Task task = new Task(description); //adds description as normal task
+        myList[listIndex] = task; //adds item to list
+        listIndex +=1;          //increases index
+
+        BinlaDan.printLine();
+        System.out.print("Added new target: ");
+        System.out.println(description); // echos what is added to list
```
I like the comments, informative

Roycecodes replied:
Thank you!

cj143n reviewed (commented):
Overall looks good and quite clean.

Roycecodes committed:
Add Text-UI-Test

Modified runtest.bat file to include correct file path

added input commands into input.txt and expected output into EXPECTED.TXT

Run runtest.bat to test

Roycecodes committed:
Improve Code Quality

Currently Many classes were located in a single .java file, this makes reading code difficult

Hence Classes were separated into their individual filed

Roycecodes committed:
Improve code quality

current code had magic numbers, redundant print statments, and non descriptive method names

changed magic numbers and reused Strings into constants, changed several method nammes to verbs that properly describe function

Roycecodes committed:
Improve code quality by adding abstract classes

Task class had redundant and confusing constructors and methods that are
no longer used, it acts solely as a parent class

Changing it into an abstract class prevents it from being accidentally
instantiated and makes the methods it holds look cleaner

additionally more magic numbers were found that were changed into
constants

Roycecodes committed:
Small fix

renamed byeText Method to make it more descriptive

Roycecodes committed:
Add exception handling for addTasks

Previously without exception handling chatbot would crash when given
inputs that do not fit the right command structure

Added try ctach statements that will allow it to fail gracefully,
keeping the chatbot running for its next command. Giving instructions on
how to better give commands

Additionally, The exception handling for done and undone is required

Roycecodes committed:
Add Exception handling for setDone and List methods

Previously code's exception handling relied on multiple if/else blocks
this is extremely inefficient as one would have to account for all the
different ways the method can error. It also makes code hard to read
because of the if else logic

Exception handling is now  handled by try catch statements and custom
exceptions like ALreadyDoneExceptions.This makes code easier to read and
understand as well as easier to write

Roycecodes committed:
Merge branch 'branch-Level-5'

Previously methods in code have no exception handling or are handled by
multiple if else statements, resulting in code failing hard when
commands are not correct and code being messy and hard to read and inefficeint because of
the multple if else statements

adding of branch-Level-5 that adds exception handling of addtask, and
set as done methods using try catch statements and custom Exceptions.
Makes code more readable and efficient

Roycecodes committed:
Add Packages and edit text UI test sources

Previously all .java files were situated in java folder making code
readability hard.

Now with packages that group similiar classes together, code is neater
and more stuctured

Additionally textUItest was edited to include all the package sources

Roycecodes committed:
Merge branch 'branch-A-Packages'

previously all java codes situated in one java file which made structure
of code messy and hard to read

Now with packages code is easier to read and classes are easier to find
chatBot can be scaled up more easily

Roycecodes committed:
Add collections

previously list was reliant on fixed array of 100 items

by implementing colelction Arraylist maintaining array is simpler and
dynamic allowing one to add more than 100 items

Roycecodes committed:
Add delete method

Previously tasks could not be deleted from list

With this method implemented tasks can be deleted from list

Roycecodes committed:
Add Saving Method

Previously all data in the session are not saved. Hence, everytime the
user opens a new session their tasks are lost.

Through the implementation of the saving method the user can now save
his data across multiple uses

Roycecodes committed:
Merge branch 'branch-Level-6'

Adding delete function

Roycecodes committed:
Merge branch-Level-7

fixed some merge conflicts between new collections

Roycecodes committed:
Fix non-existent save file error handling

previously could not create a new save file if does not exist

Roycecodes committed:
Add invalid command handling

Roycecodes committed:
main

Roycecodes committed:
Minor changes to initialise file

Roycecodes committed:
Clean up Code

previously OOP was more messy

added command class that handles commands
moved methods to more relevant classes

now code is easier to understand and read

aditichadha1310 commented on src/main/java/BinlaDan/command/Parser.java:
```diff
@@ -0,0 +1,179 @@
+        String[] returnedArray = new String[3];
+        int indexOfFrom = receivedText.indexOf("/from"); // obtaining index of first /from
+        int indexOfTo = receivedText.indexOf("/to"); // obtaining index of first /to
+        if (indexOfFrom == -1 || indexOfTo == -1) { // if any of the keywords do not exist throw missing Date exception
```
Avoid using magic numbers in the code.

aditichadha1310 reviewed (commented):
overall, good work!

Roycecodes committed:
Add LocalDate functionality

previously all dates were saved as strings
proved difficult to sort through if needed
now dates that can be parsed will be parsed
to LocalDate

this was done by using a TaskDate class

Roycecodes committed:
Add search function

Added search fucntion that helps to filter through list to find
specific tasks.

Additionally added feature where list can be modified based on
received filtered list

Roycecodes committed:
Add Java Doc

previously the lack of documentation made it hard for others to
understand code

now users can read headers to understand non obvious methods

Roycecodes committed:
Merge pull request #1 from Roycecodes/branch-Level-8

Add LocalDate functionality

Roycecodes committed:
merge master to branch

Roycecodes committed:
Merge pull request #2 from Roycecodes/branch-Level-9

Add Find function

Roycecodes committed:
Minor changes to fix Find Function

Roycecodes committed:
Merged master to branch

Roycecodes committed:
Add JavaDoc for new methods upstream merge with master

Roycecodes committed:
Merge pull request #3 from Roycecodes/branch-A-JavaDoc

Branch a java doc

Roycecodes committed:
Add delete all method

Roycecodes committed:
Add help messages

Roycecodes committed:
Minor fixes

display deleted tasks after delete all

find can filter via task type

Roycecodes committed:
Update localTaskBackups.txt

Roycecodes committed:
Add User Guide

Roycecodes committed:
Merge branch 'master' of https://github.com/Roycecodes/ip

Roycecodes committed:
Code quality checks

Roycecodes committed:
Create Jar file for release

Roycecodes committed:
corrected user guide
